crates/sim2h/src/lib.rs

```diff
@@ -1188,211 +1484,284 @@ impl Sim2h {
+                        for aspect in fetch_result.entry.aspect_list {
+                            to_remove.push((
+                                fetch_result.entry.entry_address.clone(),
+                                aspect.aspect_address.clone(),
+                            ));
+                            multi_messages.push(Lib3hToClient::HandleStoreEntryAspect(
+                                StoreEntryAspectData {
+                                    request_id: "".into(),
+                                    space_address: space_address.clone(),
+                                    provider_agent_id: agent_id.clone(),
+                                    entry_address: fetch_result.entry.entry_address.clone(),
+                                    entry_aspect: aspect,
+                                },
+                            ));
+                        }
+
+                        let store_message = WireMessage::MultiSend(multi_messages);
+
+                        let mut state = sim2h_handle.lock_state().await;
+                        let maybe_url = state.lookup_joined(&space_address, &to_agent_id);
+                        if maybe_url.is_none() {
+                            error!("Got FetchEntryResult with request id that is not a known agent id. I guess we lost that agent before we could deliver missing aspects.");
+                            return;
+                        }
+                        let url = maybe_url.unwrap();
+                        for (entry_address, aspect_address) in to_remove.drain(..) {
+                            state.remove_missing_aspect(
+                                &space_address,
+                                &to_agent_id,
+                                &entry_address,
+                                &aspect_address,
+                            );
+                        }
+                        state.send(to_agent_id, url, &store_message);
+                    });
                 }
 
                 Ok(())
             }
             WireMessage::ClientToLib3h(ClientToLib3h::QueryEntry(query_data)) => {
                 if let DhtAlgorithm::NaiveSharding {redundant_count} = self.dht_algorithm {
-                    let state = self.state.clone();
-                    let tx = self.tp_send.clone();
-                    let space_address = space_address.clone();
-                    self.threadpool.execute(move || {
-                        let disconnects =
-                            state
-                            .read()
-                            .build_query(space_address,query_data,redundant_count);
-                        tx.send(PoolTask::Disconnect(disconnects))
-                            .expect("should send");
+                    let sim2h_handle = self.sim2h_handle.clone();
+                    tokio::task::spawn(async move {
+                        let disconnects = sim2h_handle
+                            .lock_state().await
```

Comment:
Possibly good example here where a future rwlock implementation that @AshantiMutinta brought up in standup would benefit us.

Reply:
agreed :+1: - but I'm going to leave as-is for now, I think the IM swapout will obviate this